# MNIST - 서포트 벡터머신 (머신러닝 SVM)

In [1]:
import numpy as np

In [3]:
from tensorflow.keras.datasets import mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()

11501568/11490434 [==============================] - 0s 0us/step


In [4]:
X_train = X_train.reshape(-1, 784) / 255.
X_test  = X_test.reshape(-1, 784) / 255.

- SVM, GridSearchCV

In [5]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV

In [6]:
params = {'C': [0.1, 1, 5, 10]}
svc = SVC(random_state=2022)

In [8]:
# 에스티메이터는 svc, 파람그리드는 파람. 스코어링, cv 
grid_svc = GridSearchCV(
    svc, param_grid = params, scoring = 'accuracy', cv=3, n_jobs=-1
)
%time grid_svc.fit(X_train, y_train)
# 서포트 벡터 머신이라 y값 원핫인코딩할 필요없다. ??

CPU times: user 5min 28s, sys: 3.45 s, total: 5min 32s
Wall time: 1h 8min 45s


GridSearchCV(cv=3, estimator=SVC(random_state=2022), n_jobs=-1,
             param_grid={'C': [0.1, 1, 5, 10]}, scoring='accuracy')

CPU times: user 5min 28s, sys: 3.45 s, total: 5min 32s
Wall time: 1h 8min 45s
GridSearchCV(cv=3, estimator=SVC(random_state=2022), n_jobs=-1,
             param_grid={'C': [0.1, 1, 5, 10]}, scoring='accuracy')

In [10]:
grid_svc.best_params_
# {'C': 10}

{'C': 10}

In [12]:
grid_svc.best_estimator_.score(X_test, y_test)
# 0.9837

0.9837

# 세이브 중요하지

In [13]:
import joblib
joblib.dump(grid_svc.best_estimator_, 'best-mnist-svm.pkl')

['best-mnist-svm.pkl']